### DISCUSSION TITLE
Trigger a workflow when the base reference of a pull request is updated

### DISCUSSION THREAD

**Discussion by elisa-dc** [community/community#25798] [Actions]
<p>Hi,<br>
I’d like to trigger a workflow every time that there is a code change on a pull request<br>
Using the following trigger</p>
<pre><code class="lang-auto">name: run-tests
on:
  pull_request:
    branches: master
    types: [synchronize, opened, reopened, ready_for_review, unlabeled]
</code></pre>
<p>using the synchronize event, the workflow is correctly triggered every time that a new change is pushed, but not when I rebase my branch</p>
<p>Which is the correct event to use to trigger the workflow when the base reference is updated?</p>
<p>Thanks a lot!</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/elisa-dc">@elisa-dc</a>,<br>
You need to setup a workflow runs on push event in the base branch.<br>
For example, the base branch is <strong>master</strong> in your case, you can setup a workflow like as below in the <strong>master</strong> branch.</p>
<pre><code class="lang-yaml">on: push
</code></pre>
<p>OR</p>
<pre><code class="lang-yaml">on:
  push:
    branches:
      - master
</code></pre>
<p><strong>NOTE:</strong><br>
This workflow in the <strong>master</strong> branch is essentially not related to the PR, the jobs in the workflow run will not become the parts of the PR’s checks.</p>

**Comment by elisa-dc** [comment]
<p><a class="mention" href="/brightran">@brightran</a> thanks a lot for your reply.</p>
<p>I’ve tried the solution that you suggest, but it doesn’t look to be working to cover my specific use case:</p>
<ul>
<li>I have a branch x on which I’ve opened a pull request to master</li>
<li>With the solution that you suggest I can see the push event and my workflow triggered on master every time that a new change is pushed there</li>
</ul>
<p>My PR has been approved but in the meanwhile new commits have been merged on master</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/b/b15e566c153087c5f8a41586845b0c7d9cd045c4.png" data-download-href="/uploads/short-url/pj4D2fkfgmESorI4gHwtz19uKCo.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/41437937/181093333-3ad5323e-8e68-4f1d-b87e-f8d53f3253de.png" alt="image" data-base62-sha1="pj4D2fkfgmESorI4gHwtz19uKCo" width="690" height="278" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/b/b15e566c153087c5f8a41586845b0c7d9cd045c4_2_690x278.png, https://aws1.discourse-cdn.com/github/original/2X/b/b15e566c153087c5f8a41586845b0c7d9cd045c4.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/b/b15e566c153087c5f8a41586845b0c7d9cd045c4.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/b/b15e566c153087c5f8a41586845b0c7d9cd045c4_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">742×299 35.5 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Since we use a rebase strategy I might have the following situations:</p>
<ol>
<li>I rebase the branch locally and then I push =&gt; covered by the on synchronised trigger</li>
<li>I rebase triggering an Automatic rebase workflow (<a href="https://github.com/marketplace/actions/automatic-rebase" rel="nofollow noopener">https://github.com/marketplace/actions/automatic-rebase</a>) =&gt; don’t see any synchronised event</li>
<li>I rebase using the Rebase&amp;Merge button on my PR =&gt; don’t see any synchronised event</li>
</ol>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/elisa-dc">@elisa-dc</a>,</p>
<blockquote>
<p><strong>when branch x is rebased (case branch x was behind master) to a newer version of master</strong></p>
</blockquote>
<p>What is the meaning of “<strong>rebased to a newer version of master</strong>”? How do you do that?<br>
Do you mean after pushing some new commits to master, you will merge these new changes master into branch x? Any misunderstandings, feel free to tell me.</p>
<p>Here I have some suggestions may be helpful to you:</p>
<ol>
<li>Navigate to your repository’s <strong>Settings</strong> &gt; <strong>Branches</strong> &gt; <strong>Branch protection rules</strong> &gt; “<strong>Add rule</strong>” button to set protection rules for <strong>master</strong> branch.<br>
<div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/7/78cc94df2843c04a6caf3f76260be575f5b4c7ae.png" data-download-href="/uploads/short-url/heDBQe2goa4f3HCS1yBl0o6AQxw.png?dl=1" title="Branch-rules" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/46307996/181093336-49f198bb-52df-4695-a709-cdf4545ec3ce.png" alt="Branch-rules" data-base62-sha1="heDBQe2goa4f3HCS1yBl0o6AQxw" width="690" height="255" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/7/78cc94df2843c04a6caf3f76260be575f5b4c7ae_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Branch-rules</span><span class="informations">1099×407 26.4 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
</li>
<li>Enable “<strong>Require status checks to pass before merging</strong>” and “<strong>Require branches to be up to date before merging</strong>” options, and mark some checks as <strong>Required</strong>.<br>
<div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/a/a4d8ada9749b5234af18070481fb7621a4232834.png" data-download-href="/uploads/short-url/nwixfYqr2zQXS4lpFsIpcWY1ImU.png?dl=1" title="add-Branch-rules" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/46307996/181093337-11d07323-55cf-4bea-a951-6b23d4197975.png" alt="add-Branch-rules" data-base62-sha1="nwixfYqr2zQXS4lpFsIpcWY1ImU" width="640" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/a/a4d8ada9749b5234af18070481fb7621a4232834_2_640x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/a/a4d8ada9749b5234af18070481fb7621a4232834_2_960x750.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/a/a4d8ada9749b5234af18070481fb7621a4232834.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/a/a4d8ada9749b5234af18070481fb7621a4232834_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">add-Branch-rules</span><span class="informations">1077×841 50.4 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
</li>
<li>After above settings, when a PR is opened to merge into master from another branch, if push some new changes to master, then you can see the “<strong>Update branch</strong>” button appears on the PR page. Click the “<strong>Update branch</strong>” button to merge the latest changes from master into the source branch of the PR (branch x in your case).<br>
<div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/4/46dd84058bf99004e94c66a058eb31768478efbc.png" data-download-href="/uploads/short-url/a6U17LMcGgV0xF3fnT6OF7SEwQA.png?dl=1" title="Update-branch" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/46307996/181093338-599aa36e-db44-4573-af09-37076eb5439e.png" alt="Update-branch" data-base62-sha1="a6U17LMcGgV0xF3fnT6OF7SEwQA" width="690" height="456" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/4/46dd84058bf99004e94c66a058eb31768478efbc_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Update-branch</span><span class="informations">809×535 39.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
</li>
<li>Because have new changes pushed to the source branch via the “<strong>Update branch</strong>”, so the <strong>pull_request.synchronize</strong> event will trigger the workflow to run. In this way, you do not need to setup a workflow runs on push event in master branch.</li>
</ol>

**Comment by elisa-dc** [comment]
<p><a class="mention" href="/brightran">@brightran</a> thanks again for your reply. I’ve added some more information to my previous post in order to try to provide additional details on the unclear points. Using the “Update branch” indeed I see a synchronise event and the workflow triggered. Not sure if and which event is triggered in case of Rebase&amp;Merge if our repo use a rebase strategy</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/elisa-dc">@elisa-dc</a>,</p>
<p>Essentially, we can think of <strong>Rebase&amp;Merge</strong> is pushing some new commits from a branch to another branch (other merge types are similar). This is a <strong>push</strong> event for the target branch.<br>
If you setup a workflow run on push event in the target branch, after <strong>Rebase&amp;Merge</strong>, normally this workflow will be triggered.</p>
<p>Another situation, if you want a workflow triggered to run on the <strong>pull_request</strong> event after <strong>Rebase&amp;Merge</strong>, you can try using <strong>pull_request.closed</strong> event. You need to add this event to the workflow in the source branch before <strong>Rebase&amp;Merge</strong>.</p>
<pre><code class="lang-yaml">on:
  pull_request:
    types: [opened, synchronize, reopened, ready_for_review, unlabeled, closed]
    branches:
      - master
</code></pre>
<p>When a PR is merged, it is closed automatically. Then the <strong>pull_request.closed</strong> event triggers the workflow.<br>
However, you need to note that the <strong>pull_request.closed</strong> event is only evaluated for closing the PR, not for merging PR. If you manually close a PR without merged, this event still triggers the workflow.</p>

**Comment by shapeofarchitect** [comment]
<p>Hi How would you trigger a workflow when the Pull Request is merged to target branch . I want a workflow where when pull request is opened, updated, “merged” to target branch trigger a workflow. So essentially in my case I want PR merged workflow as well as the target branch (lets say develop, main) to also trigger a separate workflow when that happens.</p>
